feat(eslint-plugin): add rule [strict-void-return] (#9707)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/strict-void-return.mdx

Reviewer: > At this point this rule already did everything `no-misused-promises`'s `checkVoidReturn` did, but better. It doesn't have problems #8054 or #8739. Maybe it's worth splitting `no-misused-promises` into 3 separate rules in the future? (this one being one of them)

This is a great question. `no-misused-promises` was already largely overlapped by `no-unnecessary-condition`. This new `strict-void-return` pretty much takes on the rest of `no-misused-promises`, making `no-misused-promises` redundant if you have both `no-unnecessary-condition` and `strict-void-return`... 

I'd be in favor of deprecated `no-misused-promises` in favor of using `no-unnecessary-condition` + `strict-void-return`. The only benefit I can think of for `no-misused-promises` would be projects that want to _only_ apply the checks for Promises... Maybe these two rules could each be given some kind of _"only check Promises"_ option?

Also of note is that `no-misused-promises`'s `checkVoidReturns` is pretty configurable. Maybe, if this rule is to replace `no-misused-promises`, it'd be useful to have each of those configurable options? Or, on the other hand, maybe those options are holdovers that real-world don't generally use? Investigation needed. I think those options can be a followup & shouldn't block this PR.

What do you think?

Also cc: @typescript-eslint/triage-team in general, and @kirkwaiblinger + @alythobani from #8765.

Reviewer 2: Maybe a lame response, but is there a compelling reason not to land this first, then consider the no-misused-promises deprecation, and which options we might need to port or create in order to do so, afterwards?

Just thinking, deprecating no-misused-promises might have some strings attached, such as some nontrivial updating of the docs in no-floating-promises that explain how to lint against promise antipatterns outside of ExpressionStatements.

```ts
class Foo {
  cb() {
    console.log('foo');
  }
}

class Bar extends Foo {
  cb() {
    super.cb();
    console.log('bar');
  }
}
```

</TabItem>
</Tabs>

### Void context from implemented interfaces

This rule enforces class methods which implement a void method to also be void.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
interface Foo {
  cb(): void;
}

class Bar implements Foo {
  cb() {
    return 'cb';
  }
}
```

</TabItem>
<TabItem value="✅ Correct">

```ts
interface Foo {
  cb(): void;
}

class Bar implements Foo {
  cb() {
    console.log('cb');
  }
}
```

</TabItem>
</Tabs>

## Options

### `allowReturnAny`

{/* insert option description */}

Additional incorrect code when the option is **disabled**:

<Tabs>
<TabItem value="❌ Incorrect">

```ts option='{ "allowReturnAny": false }'
declare function fn(cb: () => void): void;

fn(() => JSON.parse('{}'));

fn(() => {
  return someUntypedApi();
});
```

</TabItem>
<TabItem value="✅ Correct">

```ts option='{ "allowReturnAny": false }'
declare function fn(cb: () => void): void;

fn(() => void JSON.parse('{}'));

fn(() => {
  someUntypedApi();
});
```

</TabItem>
</Tabs>

## When Not To Use It

> Some projects are architected so that values returned from synchronous void functions are generally safe.

Reviewer: ```suggestion
Some projects are architected so that values returned from synchronous `void` functions are generally safe.
```

> If you only want to check for misused voids with asynchronous functions then you can use [`no-misused-promises`](./no-misused-promises.mdx) instead.

Reviewer: ```suggestion
If you only want to check for misused `void`s with asynchronous functions then you can use [`no-misused-promises`](./no-misused-promises.mdx) instead.
```

> In browser context, an unhandled promise will be reported as an error in the console.

Reviewer: ```suggestion
In browser contexts, an unhandled Promise will be reported as an error in the console.
```

It's generally a good idea to also show some kind of indicator on the page that something went wrong,
but if you are just prototyping or don't care about that, the default behavior might be acceptable.
In such case, instead of handling the promises and `console.error`ing them anyways, you can just disable this rule.

> Similarly, the default behavior of crashing the process on unhandled promise rejection

Reviewer: ```suggestion
Similarly, the default server runtime behaviors of crashing the process on unhandled Promise rejections
```

:::

:::tip
Use [`no-floating-promises`](no-floating-promises.mdx)
to also enforce error handling of non-awaited promises in statement positions.
:::

### Ignored returned generators

If a generator is returned from a void function it won't even be started.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
declare function takesCallback(cb: () => void): void;

takesCallback(function* () {
  console.log('Hello');
  yield;
  console.log('World');
});
```

</TabItem>
<TabItem value="✅ Correct">

```ts
declare function takesCallback(cb: () => void): void;

takesCallback(() => {
  function* gen() {
    console.log('Hello');
    yield;
    console.log('World');
  }
  for (const _ of gen());
});
```

</TabItem>
</Tabs>

### Probable mistakes

Returning a value from a void function is likely a mistake on part of the programmer.
This rule will often warn you early when using a function in a wrong way.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
['Kazik', 'Zenek'].forEach(name => `Hello, ${name}!`);
```

</TabItem>
<TabItem value="✅ Correct">

```ts
['Kazik', 'Zenek'].forEach(name => console.log(`Hello, ${name}!`));
```

</TabItem>
</Tabs>

> ### Void context examples

Author: Removed and changed other headings slightly

> ### Void context examples

Author: I can make a blog post WIP PR directly later if you don't mind. It would contain mostly the part I removed for now.

declare let callback: () => void;
declare let returnsCallback: () => () => void;
declare let callbackObj: Record<string, () => void>;
declare let callbackArr: (() => void)[];

takesCallback(() => console.log('Hello'));
<TakesCallback cb={() => console.log('Hello')} />;
callback = () => console.log('Hello');
returnsCallback = () => {
  return () => console.log('Hello');
};
callbackObj = {
  hello: () => console.log('Hello'),
};
callbackArr = [() => console.log('Hello')];
```

</TabItem>
</Tabs>

### Void context from overloads

This rule treats an any-returning callback argument as a void context,
if there exists another overload where it is typed as returning void.

This is required to correctly detect `addEventListener`'s callback as void callback,
because otherwise the call always resolves to the any-returning signature.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
/// <reference lib="dom" />

document.addEventListener('click', () => {
  return 'Clicked';
});
```

</TabItem>
<TabItem value="✅ Correct">

```ts
/// <reference lib="dom" />

document.addEventListener('click', () => {
  console.log('Clicked');
});
```

</TabItem>
</Tabs>

> ### Void context from base class

Author: I already changed it but I just remembered my reasoning: You can only extend one class in a class declaration (unlike many interfaces) that's why this one was singular :laughing:

---
description: 'Disallow passing a value-returning function in a position accepting a void function.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/strict-void-return** for documentation.

## Rule Details

TypeScript considers functions returning a value to be assignable to a function returning void.
Using this feature of TypeScript can lead to bugs or confusing code.

## Examples

### Unsafety

Passing a value-returning function in a place expecting a void function can be unsound.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
const bad: () => void = () => 2137;

> const func = Math.random() > 0.1 ? bad : prompt;

Author: I tried

---
description: 'Disallow passing a value-returning function in a position accepting a void function.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/strict-void-return** for documentation.

## Rule Details

TypeScript considers functions returning a value to be assignable to a function returning void.
Using this feature of TypeScript can lead to bugs or confusing code.

## Examples

### Return type unsafety

> Passing a value-returning function in a place expecting a void function can be unsound.

Reviewer 2: What do you think of a little more verbosity here? 

If I were writing it, I might do something like this...

````suggestion
Passing a value-returning function in a place expecting a void function can be unsound. TypeScript generally treats the `void` type as though it has the same runtime behavior as `undefined`, so this pattern will cause a mismatch between the runtime behavior and the types.

```ts
// TypeScript errors on overtly wrong ways of populating the `void` type...
function returnsVoid(): void {
    return 'not void!'; // TS Error:  Type 'string' is not assignable to type 'void'.
}

// ...but allows more subtle ones
const returnsVoid: () => void = () => 1234;

// Likewise, TypeScript errors on overtly wrong usages of `void` as a runtime value...
declare const v: void;
if (v) { // TS Error: An expression of type 'void' cannot be tested for truthiness.
    // ... do something
}

// ...but silently treats `void` as `undefined` in more subtle scenarios
declare const voidOrString: void | string;
if (voidOrString) {
     // voidOrString is narrowed to string in this branch, so this is allowed.
     console.log(voidOrString.toUpperCase());
}
```

Putting these behaviors together, we can construct a scenario where TS will throw at runtime, despite not reporting a type error:

````

<Tabs>
<TabItem value="❌ Incorrect">

```ts
const getNothing: () => void = () => 2137;
const getString: () => string = () => 'Hello';

> const getStringMaybe = Math.random() > 0.1 ? getNothing : getString;
> const maybeString = getStringMaybe();

Reviewer 2: (praise) This version of the example is much easier to follow. Thanks! 🙂 

(nit) why not inline the `getStringMaybe` variable still?
```suggestion
const maybeString = Math.random() > 0.1 ? getNothing() : getString();
```
